lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -1131,7 +1131,48 @@ TypeExpr *PreCheckExpression::simplifyTypeExpr(Expr *E) {
                                         ResultTypeRepr);
     return new (TC.Context) TypeExpr(TypeLoc(NewTypeRepr, Type()));
   }
-
+  
+  // Fold 'P & Q' into a composition type
+  if (auto *binaryExpr = dyn_cast<BinaryExpr>(E)) {
+    auto Operator = Lexer::getTokenAtLocation(TC.Context.SourceMgr,
+                                              binaryExpr->getLoc());
+    if (Operator.isContextualPunctuator("&")) {
+      // The protocols we are composing
+      SmallVector<IdentTypeRepr *, 4> Protocols;
+      
+      auto lhsExpr = binaryExpr->getArg()->getElement(0);
+      if (auto *lhs = dyn_cast_or_null<TypeExpr>(lhsExpr)) {
+        if (auto repr = dyn_cast<IdentTypeRepr>(lhs->getTypeRepr()))
+          Protocols.push_back(repr);
+      }
+      // If the lhs is another binary expression, we have a multi element composition:
+      // 'A & B & C' is parsed as ((A & B) & C), so we get the protocols from the lhs here
+      else if (isa<BinaryExpr>(lhsExpr)) {
```

Comment:
This stretch can probably be structured more gracefully by inverting the condition and returning `nullptr` once.